File: legalapp-plotchoice/Landed-legal/server/src/services/calculationEngine.ts

export interface StampDutyInput {
  instrumentType: string;
  considerationValue: number;
  guidelineValue: number;
  isFemaleBuyer?: boolean;
}

export interface StampDutyOutput {
  taxableMarketValue: number;
  stampDutyPercent: number;
  stampDutyAmount: number;
  registrationFeePercent: number;
  registrationFeeAmount: number;
  surchargePercent: number;
  surchargeAmount: number;
  totalOutlay: number;
  calculationBasis: string;
}

export function calculateStampDuty(input: StampDutyInput): StampDutyOutput {
  const taxableValue = Math.max(input.considerationValue || 0, input.guidelineValue || 0);
  
  let stampDutyPct = 7;
  let regFeePct = 2;
  let surchargePct = 0;

  switch (input.instrumentType?.toLowerCase()) {
    case 'gift':
      stampDutyPct = 7;
      regFeePct = 2;
      break;
    case 'settlement_family':
    case 'settlement':
      stampDutyPct = 1;
      regFeePct = 1;
      break;
    case 'partition':
      stampDutyPct = 1;
      regFeePct = 1;
      break;
    case 'release':
      stampDutyPct = 1;
      regFeePct = 1;
      break;
    case 'mortgage':
      stampDutyPct = 1;
      regFeePct = 1;
      break;
    case 'lease':
      stampDutyPct = 1;
      regFeePct = 1;
      break;
    case 'sale':
    default:
      stampDutyPct = 7;
      regFeePct = 2;
      surchargePct = 0;
      break;
  }

  const stampDutyAmt = Math.round((taxableValue * stampDutyPct) / 100);
  const regFeeAmt = Math.round((taxableValue * regFeePct) / 100);
  const surchargeAmt = Math.round((taxableValue * surchargePct) / 100);
  const total = stampDutyAmt + regFeeAmt + surchargeAmt;

  return {
    taxableMarketValue: taxableValue,
    stampDutyPercent: stampDutyPct,
    stampDutyAmount: stampDutyAmt,
    registrationFeePercent: regFeePct,
    registrationFeeAmount: regFeeAmt,
    surchargePercent: surchargePct,
    surchargeAmount: surchargeAmt,
    totalOutlay: total,
    calculationBasis: input.considerationValue > input.guidelineValue
      ? 'Calculated on Agreed Consideration Value (higher than Guideline Value)'
      : 'Calculated on Government Guideline Value (higher than Agreed Consideration)'
  };
}

export interface BuildingValueParams {
  buildingType: string;
  ageYears: number;
  totalAreaSqFt: number;
  floorsCount?: number;
  amenitiesCost?: number;
}

export function calculateBuildingValue(params: BuildingValueParams) {
  let baseRateSqFt = 1800; // PWD RCC Standard rate
  if (params.buildingType?.toLowerCase().includes('commercial')) baseRateSqFt = 2400;
  if (params.buildingType?.toLowerCase().includes('industrial')) baseRateSqFt = 1500;
  if (params.buildingType?.toLowerCase().includes('tiled')) baseRateSqFt = 1100;

  const grossReplacementValue = params.totalAreaSqFt * baseRateSqFt;
  
  // Standard PWD depreciation: 1.5% per annum capped at 60%
  const depreciationPct = Math.min(params.ageYears * 1.5, 60);
  const depreciationAmount = Math.round((grossReplacementValue * depreciationPct) / 100);
  const depreciatedStructureValue = grossReplacementValue - depreciationAmount;
  const totalAmenities = params.amenitiesCost || 0;
  const netEstimatedValue = depreciatedStructureValue + totalAmenities;

  return {
    totalAreaSqFt: params.totalAreaSqFt,
    basePlinthRatePerSqFt: baseRateSqFt,
    grossReplacementValue,
    depreciationPercent: depreciationPct,
    depreciationAmount,
    depreciatedStructureValue,
    amenitiesCost: totalAmenities,
    estimatedBuildingValue: netEstimatedValue
  };
}

File: legalapp-plotchoice/Landed-legal/fastapi_backend/app/core/calculation_engine.py
def calculate_stamp_duty(instrument_type: str, consideration_value: float, guideline_value: float, is_female_buyer: bool = False):
    taxable_value = max(consideration_value or 0.0, guideline_value or 0.0)
    
    inst = (instrument_type or "sale").lower()
    stamp_duty_pct = 7.0
    reg_fee_pct = 2.0
    surcharge_pct = 0.0

    if inst in ["settlement", "settlement_family", "partition", "release", "mortgage", "lease"]:
        stamp_duty_pct = 1.0
        reg_fee_pct = 1.0
    elif inst == "gift":
        stamp_duty_pct = 7.0
        reg_fee_pct = 2.0

    stamp_duty_amt = round((taxable_value * stamp_duty_pct) / 100.0)
    reg_fee_amt = round((taxable_value * reg_fee_pct) / 100.0)
    surcharge_amt = round((taxable_value * surcharge_pct) / 100.0)
    total_outlay = stamp_duty_amt + reg_fee_amt + surcharge_amt

    return {
        "taxableMarketValue": taxable_value,
        "stampDutyPercent": stamp_duty_pct,
        "stampDutyAmount": stamp_duty_amt,
        "registrationFeePercent": reg_fee_pct,
        "registrationFeeAmount": reg_fee_amt,
        "surchargePercent": surcharge_pct,
        "surchargeAmount": surcharge_amt,
        "totalOutlay": total_outlay,
        "calculationBasis": "Calculated on Agreed Consideration Value (higher than Guideline Value)" if consideration_value > guideline_value else "Calculated on Government Guideline Value (higher than Agreed Consideration)"
    }

def calculate_building_value(building_type: str, age_years: int, total_area_sqft: float, amenities_cost: float = 0.0):
    base_rate = 1800.0
    b_type = (building_type or "residential_rcc").lower()
    if "commercial" in b_type:
        base_rate = 2400.0
    elif "industrial" in b_type:
        base_rate = 1500.0
    elif "tiled" in b_type:
        base_rate = 1100.0

    gross_replacement_value = total_area_sqft * base_rate
    depreciation_pct = min(age_years * 1.5, 60.0)
    depreciation_amount = round((gross_replacement_value * depreciation_pct) / 100.0)
    depreciated_structure_value = gross_replacement_value - depreciation_amount
    net_estimated_value = depreciated_structure_value + (amenities_cost or 0.0)

    return {
        "totalAreaSqFt": total_area_sqft,
        "basePlinthRatePerSqFt": base_rate,
        "grossReplacementValue": gross_replacement_value,
        "depreciationPercent": depreciation_pct,
        "depreciationAmount": depreciation_amount,
        "depreciatedStructureValue": depreciated_structure_value,
        "amenitiesCost": amenities_cost or 0.0,
        "estimatedBuildingValue": net_estimated_value
    }